File: cpp/14_abstraction.cpp
// ============================================================
// Topic:  Abstraction
// Summary: A pure-virtual method makes a class abstract (cannot instantiate).
// Run:     g++ -std=c++17 cpp/14_abstraction.cpp -o cpp/14_abstraction && cpp/14_abstraction
// ============================================================

#include <iostream>
#include <iomanip>
#include <sstream>
#include <string>

const double PI = 3.141592653589793;

// ----- 1. Abstract class: a "= 0" pure virtual declares WHAT, not HOW -----
class Shape {
public:
    virtual ~Shape() = default;
    virtual double area() const = 0;        // pure virtual -> Shape is abstract
    virtual std::string label() const = 0;

    // concrete method built on the abstract ones
    std::string describe() const {
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(2);
        oss << label() << " has area " << area();
        return oss.str();
    }
};

// ----- 2. Concrete subclasses MUST override every pure virtual -----
class Circle : public Shape {
    int r;
public:
    Circle(int r) : r(r) {}
    double area() const override { return PI * r * r; }
    std::string label() const override { return "Circle with radius " + std::to_string(r); }
};

class Square : public Shape {
    int s;
public:
    Square(int s) : s(s) {}
    double area() const override { return static_cast<double>(s * s); }
    std::string label() const override { return "Square with side " + std::to_string(s); }
};

int main() {
    // Shape s; // error: cannot declare variable 's' to be of abstract type 'Shape'
    std::cout << Circle(2).describe() << std::endl; // Circle with radius 2 has area 12.57
    std::cout << Square(3).describe() << std::endl; // Square with side 3 has area 9.00
    return 0;
}

// Cross-language note:
//   C++ has no `abstract` keyword — a class becomes abstract simply by having a
//   pure virtual method (`= 0`). Such a class can't be instantiated; a subclass that
//   leaves any pure virtual unimplemented stays abstract too.

File: rust/14_abstraction.rs
// ============================================================
// Topic:  Abstraction  (Rust: a trait IS the abstraction/interface)
// Summary: A trait defines required methods + a default; you can't instantiate it.
// Run:     rustc rust/14_abstraction.rs -o rust/14_abstraction && rust/14_abstraction
// ============================================================

// ----- 1. A trait declares the contract: WHAT, not HOW -----
trait Shape {
    fn area(&self) -> f64;        // required method (no body)
    fn label(&self) -> String;    // required method

    // default method = the concrete, shared part of the contract
    fn describe(&self) -> String {
        format!("{} has area {:.2}", self.label(), self.area())
    }
}

// ----- 2. Each type implements the trait's required methods -----
struct Circle {
    r: i32,
}
impl Shape for Circle {
    fn area(&self) -> f64 { std::f64::consts::PI * (self.r * self.r) as f64 }
    fn label(&self) -> String { format!("Circle with radius {}", self.r) }
}

struct Square {
    s: i32,
}
impl Shape for Square {
    fn area(&self) -> f64 { (self.s * self.s) as f64 }
    fn label(&self) -> String { format!("Square with side {}", self.s) }
}

fn main() {
    // let s = Shape; // ERROR: a trait is not a type you can construct
    println!("{}", Circle { r: 2 }.describe()); // Output: Circle with radius 2 has area 12.57
    println!("{}", Square { s: 3 }.describe()); // Output: Square with side 3 has area 9.00
}

// Cross-language note:
//   Rust has no `abstract class` — a TRAIT is both the interface (required methods)
//   and the place for shared default methods. A trait isn't a concrete type, so you
//   can't instantiate it directly; you implement it for a struct and use that.
